// 100% auto-generated by Source Genesys WebApi Front & Back Creator
// copyright © 2000-2025 Menphis - Sistemas Inteligentes
// This file is part of the Source Genesys project                     
'use client';
import { CRUD_CONSTANTS } from '@/app/tools/crud';
import { InstanciaApi, InstanciaApiError } from '../Apis/ApiInstancia';
import { FilterInstancia } from '../Filters/Instancia';
import { IInstancia } from '../Interfaces/interface.Instancia';
import { InstanciaEmpty } from '../../Models/Instancia';

export class InstanciaValidator {
  static validateInstancia(instancia: IInstancia): { isValid: boolean; errors: string[] } {
    const errors: string[] = [];

    // Atualmente não há validações de regras de negócio específicas
    // Todas as validações são feitas nos inputs correspondentes
    
    return {
      isValid: errors.length === 0,
      errors
    };
  }
}

export interface IInstanciaService {
  fetchInstanciaById: (id: number) => Promise<IInstancia>;
  saveInstancia: (instancia: IInstancia) => Promise<IInstancia>;  
  getList: (filtro?: FilterInstancia) => Promise<IInstancia[]>;
  getAll: (filtro?: FilterInstancia) => Promise<IInstancia[]>;
  deleteInstancia: (id: number) => Promise<void>;
  validateInstancia: (instancia: IInstancia) => { isValid: boolean; errors: string[] };
}

export class InstanciaService implements IInstanciaService {
  constructor(private api: InstanciaApi) {}

  async fetchInstanciaById(id: number): Promise<IInstancia> {
    if (id <= 0) {
      throw new InstanciaApiError('ID inválido', 400, 'INVALID_ID');
    }

    try {
      
      const response = await this.api.getById(id);
      return response.data;

    } catch (error) {
      if (error instanceof InstanciaApiError) {
        throw error;
      }
      throw new InstanciaApiError('Erro ao buscar instancia', 500, 'FETCH_ERROR', error);
    }
  }

  async saveInstancia(instancia: IInstancia): Promise<IInstancia> {    
    const validation = this.validateInstancia(instancia);
    if (!validation.isValid) {
      throw new InstanciaApiError(
        `Dados inválidos: ${validation.errors.join(', ')}`,
        400,
        'VALIDATION_ERROR'
      );
    }

    try {
      const response = await this.api.addAndUpdate(instancia);
      return response.data;
    } catch (error) {
      if (error instanceof InstanciaApiError) {
        throw error;
      }
      throw new InstanciaApiError('Erro ao salvar instancia', 500, 'SAVE_ERROR', error);
    }
  }

  
    async getList(filtro?: FilterInstancia): Promise<IInstancia[]> {
    try {
      const response = await this.api.getListN(CRUD_CONSTANTS.MAX_RECORDS_COMBO, filtro);
      return response.data || [];
    } catch (error) {
      console.log('Error fetching Instancia list');
      return [];
    }
  }

 
  

   async getAll(
    filtro?: FilterInstancia,
    onOnlineData?: (data: IInstancia[]) => void
  ): Promise<IInstancia[]> {
    try {
      // Carrega dados offline primeiro
      const preloadResponse = await this.api.filterPreload(0, filtro ?? {});
      const offlineData = preloadResponse?.data || [];

      if (onOnlineData) {
        // Busca dados online em background e envia via callback
        this.api.filter(0, filtro ?? {})
          .then(response => {
            if (response?.data) {
              onOnlineData(response.data);
            }
          })
          .catch(error => {
            console.log('Error fetching online Instancia');
          });
        // Retorna offline imediatamente
        return offlineData;
      } else {
        // Se não há callback, aguarda dados online e retorna
        try {
          const onlineResponse = await this.api.filter(0, filtro ?? {});
          return onlineResponse?.data || offlineData;
        } catch (error) {
          console.log('Error fetching online Instancia');
          return offlineData;
        }
      }
    } catch (error) {
      console.log('Error fetching all Instancia:');
      return [];
    }
  }


  async deleteInstancia(id: number): Promise<void> {
    if (id <= 0) {
      throw new InstanciaApiError('ID inválido para exclusão', 400, 'INVALID_ID');
    }

    try {
      await this.api.delete(id);
    } catch (error) {
      if (error instanceof InstanciaApiError) {
        throw error;
      }
      throw new InstanciaApiError('Erro ao excluir instancia', 500, 'DELETE_ERROR', error);
    }
  }

  validateInstancia(instancia: IInstancia): { isValid: boolean; errors: string[] } {
    return InstanciaValidator.validateInstancia(instancia);
  }
}